File: mcp_server/memory/skill_scan.py
"""Auto skill scan: proposes skills from cross-project lesson patterns.

Runs opportunistically inside briefing() on a time gate (the MCP server has
no background scheduler), looking for two things:

1. Domains with no compiled skill whose episodic pool already carries rules
   that would clear the reinforcement gate — by default only when a rule
   spans more than one project, since a lesson that recurs across projects
   is the strongest signal it deserves to become policy.
2. Existing skills whose pending changes (as detected by the briefing's
   update-diff surface) are worth turning into a concrete draft.

Everything it produces is a proposal stash — exactly what a human running
compile_skill(mode="propose") would create — so the propose-and-accept gate
is untouched: nothing here ever writes a skill. Noise control is a per-domain
"seen" marker holding the sha of the last auto-proposed body; a draft the
human let expire is not re-proposed until the compiled output actually
changes, so ignoring a proposal is a valid way to decline it.
"""

import logging
import os
import time
from typing import Any, Iterable

from .skill_compiler import compile_skill_flow, proposal_key
from .skills import (
    GENERATED_SKILL_PREFIX,
    body_sha,
    build_rules,
    extract_lessons,
    gather_domain_pools,
    known_domains,
    lesson_bearing,
    skill_user,
    strip_volatile,
    validate_domain,
)

logger = logging.getLogger(__name__)

_LAST_RUN_KEY = "meta:skill_scan:last_run"
_SEEN_PREFIX = "meta:skill_scan:seen:"
_PROPOSAL_PREFIX = "meta:skill:proposal:"

# New-skill discovery runs the full extract/cluster pipeline per candidate,
# so the candidate list is capped (largest pools first) to bound the work.
_MAX_CANDIDATES = 10


def scan_due(store) -> bool:
    """True when the time gate has elapsed. SKILL_SCAN_INTERVAL_HOURS=0 disables."""
    try:
        hours = float(os.getenv("SKILL_SCAN_INTERVAL_HOURS", "24"))
    except ValueError:
        hours = 24.0
    if hours <= 0:
        return False
    raw = store.client.get(_LAST_RUN_KEY)
    if raw:
        try:
            if time.time() - float(raw) < hours * 3600:
                return False
        except (TypeError, ValueError):
            pass
    return True


def _pending_proposal_domains(store) -> set[str]:
    """Domains with a live proposal stash — human or auto, hands off either way."""
    keys = store.scan_prefix(_PROPOSAL_PREFIX)
    if not keys:
        return set()
    rows = store.get_fields_multi(keys, ("domain",))
    return {row["domain"] for row in rows if row and row.get("domain")}


def _existing_skill_domains(store) -> set[str]:
    keys = store.scan_prefix(GENERATED_SKILL_PREFIX)
    if not keys:
        return set()
    rows = store.get_fields_multi(keys, ("domain",))
    return {row["domain"] for row in rows if row and row.get("domain")}


def _propose(
    store, embedder, domain: str, user: str, new_skill: bool,
) -> dict[str, Any] | None:
    """Run the shared propose flow and apply the seen-sha noise gate.

    Returns a briefing entry, or None when the domain didn't produce a
    proposal or produced the same draft the human already ignored.
    """
    result = compile_skill_flow(store, embedder, domain, mode="propose")
    if result.get("status") != "proposal":
        return None

    stash = store.get(proposal_key(domain, user))
    body = (stash or {}).get("body", "")
    sha = body_sha(strip_volatile(body))
    seen_key = f"{_SEEN_PREFIX}{domain}-{user}"
    if store.client.get(seen_key) == sha:
        # Same draft as last time and it was left to expire — declining by
        # ignoring is valid, so withdraw the stash and stay quiet until the
        # compiled output actually changes.
        store.delete(proposal_key(domain, user))
        return None
    store.client.set(seen_key, sha)

    entry: dict[str, Any] = {
        "domain": domain,
        "skill_id": result.get("skill_id", ""),
        "new_skill": new_skill,
        "review": f"compile_skill(domain='{domain}', mode='propose') to see the "
                  f"{'draft' if new_skill else 'diff'}, mode='write' to accept",
    }
    if result.get("rules"):
        entry["rules"] = result["rules"]
    if not new_skill and result.get("changes"):
        entry["changes"] = result["changes"]
    return entry


def run_skill_scan(
    store, embedder, update_domains: Iterable[str] = (),
) -> dict[str, Any]:
    """One scan pass: propose new skills, then drafts for changed skills.

    update_domains: domains of existing skills the caller already knows have
    pending changes (the briefing passes its pending_skill_updates result in,
    so change detection isn't run twice).
    """
    now = time.time()
    # Stamp first so a failing scan waits out the interval instead of
    # retrying on every briefing.
    store.client.set(_LAST_RUN_KEY, str(now))

    max_proposals = max(0, int(os.getenv("SKILL_SCAN_MAX_PROPOSALS", "3")))
    min_pool = max(1, int(os.getenv("SKILL_SCAN_MIN_POOL", "3")))
    cross_project = (
        os.getenv("SKILL_SCAN_CROSS_PROJECT", "true").strip().lower()
        not in ("false", "0", "no")
    )

    user = skill_user()
    existing = _existing_skill_domains(store)
    pending = _pending_proposal_domains(store)
    proposals: list[dict[str, Any]] = []
    candidates_checked = 0

    # Phase 1: domains with no skill yet, largest pools first.
    counts = known_domains(store)
    candidates: list[str] = []
    for domain in sorted(counts, key=lambda d: (-counts[d], d)):
        if domain in existing or domain in pending:
            continue
        try:
            validate_domain(domain)
        except ValueError:
            continue
        candidates.append(domain)
        if len(candidates) >= _MAX_CANDIDATES:
            break

    pools = gather_domain_pools(store, candidates) if candidates else {}
    for domain in candidates:
        if len(proposals) >= max_proposals:
            break
        pool = pools.get(domain, [])
        if sum(1 for m in pool if lesson_bearing(m)) < min_pool:
            continue
        candidates_checked += 1
        eligible, _ = build_rules(embedder, extract_lessons(pool))
        if not eligible:
            continue
        if cross_project and not any(len(r.projects) >= 2 for r in eligible):
            continue
        entry = _propose(store, embedder, domain, user, new_skill=True)
        if entry:
            proposals.append(entry)

    # Phase 2: drafts for existing skills the briefing flagged as changed.
    for domain in update_domains:
        if len(proposals) >= max_proposals:
            break
        if domain in pending or domain not in existing:
            continue
        entry = _propose(store, embedder, domain, user, new_skill=False)
        if entry:
            proposals.append(entry)

    if proposals:
        logger.info(
            "Skill scan proposed %d draft(s): %s",
            len(proposals), ", ".join(p["domain"] for p in proposals),
        )
    return {
        "ran_at": str(now),
        "proposals": proposals,
        "new_skill_candidates_checked": candidates_checked,
    }

File: crates/omnimem-engine/src/skill_scan.rs
//! The auto skill scan (`memory/skill_scan.py`).
//!
//! Time-gated inside `briefing()`, it proposes skills for domains whose
//! lessons already clear the gate (by default only when a rule spans more
//! than one project) and drafts for skills with pending changes. Everything
//! it produces is a proposal stash, so the propose-and-accept gate is
//! untouched. A per-domain "seen" marker holds the sha of the last
//! auto-proposed body, so a draft left to expire is not proposed again until
//! the compiled output changes: ignoring a proposal declines it.

use std::collections::HashSet;

use serde_json::{Map, Value, json};
use tracing::info;

use crate::compiler::proposal_key;
use crate::domains::is_valid_domain;
use crate::pyfmt::{now_secs, py_float};
use crate::skills::{
    GENERATED_SKILL_PREFIX, body_sha, extract_lessons, lesson_bearing, strip_volatile,
};
use crate::{Engine, Result};

const LAST_RUN_KEY: &str = "meta:skill_scan:last_run";
const SEEN_PREFIX: &str = "meta:skill_scan:seen:";
const PROPOSAL_PREFIX: &str = "meta:skill:proposal:";

/// Candidates are capped, largest pools first, to bound the work.
const MAX_CANDIDATES: usize = 10;

/// What one project declaring a domain adds to its candidate ranking.
const PROJECT_DOMAIN_WEIGHT: usize = 2;

impl Engine {
    /// True when the time gate has elapsed.
    pub(crate) fn skill_scan_due(&self) -> Result<bool> {
        let hours = self.config.skill_scan_interval_hours;
        if hours <= 0.0 {
            return Ok(false);
        }
        if let Some(last) = self
            .store
            .string_get(LAST_RUN_KEY)?
            .and_then(|raw| raw.trim().parse::<f64>().ok())
            && now_secs() - last < hours * 3600.0
        {
            return Ok(false);
        }
        Ok(true)
    }

    /// Domains with a live proposal, human or auto: hands off either way.
    fn pending_proposal_domains(&self) -> Result<HashSet<String>> {
        let mut domains = HashSet::new();
        for key in self.store.scan_prefix(PROPOSAL_PREFIX)? {
            if let Some(domain) = self
                .store
                .hash_get_all(&key)?
                .and_then(|f| f.get("domain").cloned())
                .filter(|d| !d.is_empty())
            {
                domains.insert(domain);
            }
        }
        Ok(domains)
    }

    fn existing_skill_domains(&self) -> Result<HashSet<String>> {
        let keys = self.store.scan_prefix(GENERATED_SKILL_PREFIX)?;
        Ok(self
            .store
            .get_fields_multi(&keys, &["domain"])?
            .into_iter()
            .flatten()
            .filter_map(|r| r.get("domain").cloned().filter(|d| !d.is_empty()))
            .collect())
    }

    /// Run the shared propose flow and apply the seen-sha noise gate.
    fn scan_propose(&self, domain: &str, new_skill: bool) -> Result<Option<Value>> {
        let result = self.compile_skill(domain, "propose", 2, true, None, None)?;
        if result["status"] != "proposal" {
            return Ok(None);
        }
        let user = &self.config.skill_user;
        let key = proposal_key(domain, user);
        let body = self
            .store
            .hash_get_all(&key)?
            .and_then(|s| s.get("body").cloned())
            .unwrap_or_default();
        let sha = body_sha(&strip_volatile(&body));
        let seen_key = format!("{SEEN_PREFIX}{domain}-{user}");
        if self.store.string_get(&seen_key)?.as_deref() == Some(sha.as_str()) {
            // The same draft the human already let expire: withdraw it.
            self.store.kv_delete(&key)?;
            return Ok(None);
        }
        self.store.string_set(&seen_key, &sha, None)?;

        let mut entry = Map::new();
        entry.insert("domain".into(), domain.into());
        entry.insert("skill_id".into(), result["skill_id"].clone());
        entry.insert("new_skill".into(), new_skill.into());
        entry.insert(
            "review".into(),
            format!(
                "compile_skill(domain='{domain}', mode='propose') to see the {}, mode='write' to accept",
                if new_skill { "draft" } else { "diff" }
            )
            .into(),
        );
        if let Some(rules) = result
            .get("rules")
            .filter(|r| r.as_object().is_some_and(|o| !o.is_empty()))
        {
            entry.insert("rules".into(), rules.clone());
        }
        if !new_skill && let Some(changes) = result.get("changes") {
            entry.insert("changes".into(), changes.clone());
        }
        Ok(Some(Value::Object(entry)))
    }

    /// One scan pass. `update_domains` are skills the briefing already knows
    /// have pending changes.
    pub(crate) fn run_skill_scan(&self, update_domains: &[String]) -> Result<Value> {
        let now = now_secs();
        // Stamp first, so a failing scan waits out the interval.
        self.store.string_set(LAST_RUN_KEY, &py_float(now), None)?;

        let max_proposals = self.config.skill_scan_max_proposals.max(0) as usize;
        let min_pool = self.config.skill_scan_min_pool.max(1) as usize;
        let existing = self.existing_skill_domains()?;
        let pending = self.pending_proposal_domains()?;
        let mut proposals: Vec<Value> = Vec::new();
        let mut checked = 0;

        let mut counts = self.known_domains()?;
        // Boost, never introduce: a domain nothing is tagged with would be
        // dropped at the pool gate anyway.
        let project_domains = self.domain_map()?;
        for (domain, count) in &mut counts {
            if let Some(projects) = project_domains.get(domain) {
                *count += PROJECT_DOMAIN_WEIGHT * projects.len();
            }
        }
        counts.sort_by(|a, b| b.1.cmp(&a.1).then_with(|| a.0.cmp(&b.0)));
        let candidates: Vec<String> = counts
            .into_iter()
            .map(|(d, _)| d)
            .filter(|d| !existing.contains(d) && !pending.contains(d) && is_valid_domain(d))
            .take(MAX_CANDIDATES)
            .collect();

        let pools = self.gather_domain_pools(&candidates)?;
        for domain in &candidates {
            if proposals.len() >= max_proposals {
                break;
            }
            let pool = pools.get(domain).map(Vec::as_slice).unwrap_or_default();
            if pool.iter().filter(|m| lesson_bearing(m)).count() < min_pool {
                continue;
            }
            checked += 1;
            let (eligible, _) = self.build_rules(&extract_lessons(pool, true), 2)?;
            if eligible.is_empty() {
                continue;
            }
            if self.config.skill_scan_cross_project
                && !eligible.iter().any(|r| r.projects.len() >= 2)
            {
                continue;
            }
            if let Some(entry) = self.scan_propose(domain, true)? {
                proposals.push(entry);
            }
        }

        for domain in update_domains {
            if proposals.len() >= max_proposals {
                break;
            }
            if pending.contains(domain) || !existing.contains(domain) {
                continue;
            }
            if let Some(entry) = self.scan_propose(domain, false)? {
                proposals.push(entry);
            }
        }

        if !proposals.is_empty() {
            let domains: Vec<&str> = proposals
                .iter()
                .filter_map(|p| p["domain"].as_str())
                .collect();
            info!(count = proposals.len(), domains = %domains.join(", "), "skill scan proposed drafts");
        }
        Ok(json!({
            "ran_at": py_float(now),
            "proposals": proposals,
            "new_skill_candidates_checked": checked,
        }))
    }
}
